ADR/nnnn-trusted-task-model.md

```diff
@@ -0,0 +1,245 @@
+
+Scaling this approach to a decentralized setting, where teams maintain separate
+Task repositories, is problematic. Currently, the Konflux release pipeline for Task
+bundles updates the `data-acceptable-bundles` artifact. But the update mechanism
+was not designed to handle parallel updates. When multiple releases run in parallel,
+the simultaneous updates can corrupt the content of `data-acceptable-bundles`. The
+original plan to avoid this problem was that there would be many "acceptable bundles"
+artifacts, one for each repository. But that doesn't solve the problem. Each Task
+bundle is a separate Konflux component, a repository can contain many Task bundles,
+their releases can run in parallel.
+
+Konflux needs a new model for trusting Tasks after decentralization.
+
+## Decision
+
+### Adopt a rule-based model for Trusted Tasks
+
+In addition to the `trusted_tasks` data, Conforma will gain the ability to trust
+Tasks based on `trusted_task_rules`. The rules will give policy authors the ability
+to allow/deny sets of Tasks based on some criteria.
+
+If the current logic for trusting Tasks is as follows (pseudocode):
+
+```text
+is_on_trusted_tasks_list(task_reference)
+and not expired_on_trusted_tasks_list(task_reference)
+```
+
+Then, with the addition of `trusted_task_rules`, it becomes:
+
+```text
+(
+  is_on_trusted_tasks_list(task_reference)
+  or is_allowed_by_trusted_task_rules(task_reference)
+)
+and not expired_on_trusted_tasks_list(task_reference)
+and not is_denied_by_trusted_task_rules(task_reference)
+```
+
+Note:
+
+- Both the `trusted_tasks` and the `trusted_task_rules` mechanisms should be supported,
+  at least until decentralization is complete
+- Denial takes priority
+  - Even if a Task is on the `trusted_tasks` list, a deny rule should be able to
+    make it untrusted.
+  - Even if an allow rule makes a Task trusted, if it's expired according to the
+    `trusted_tasks` list, it should be considered untrusted.
+
+This ADR doesn't mandate a specific format for `trusted_task_rules`, but does
+come with an example of what it could look like - see the [Appendix](#appendix).
+
+### Trust upstream Konflux Tasks based on bundle URL
+
+A policy author who wants to trust all the upstream Konflux Tasks should do so by
+defining a rule that allows all task bundles from `quay.io/konflux-ci/tekton-catalog/*`
+with no further conditions. The `trusted_task_rules` mechanism must support this.
+
+### Start signing upstream Konflux tasks with a common release key
+
+Trusting all task bundles from `quay.io/konflux-ci/tekton-catalog/*` is not a strong
+trust model (but isn't weaker than the current one - see [Consequences](#consequences)).
+To let policy authors define a stronger trust model, we should aim to start signing
+the upstream Konflux task bundles with a common release key and extend the
+`trusted_task_rules` mechanism to support signature verification.
```

Comment:
Should we have specifics about how to establish trust in these tasks as a minimum suggested gate? This has some relevancy with the partner task ADR as well as @ralphbean already mentioned.

Reply:
Do you mean rules for letting tasks get released to `quay.io/konflux-ci/tekton-catalog`?

We probably should have those, but I don't know how to go about it. It seems like that is up to the reviewers in the konflux-release-data repo in an internal gitlab instance :/

Reply:
It would be related to the conforma policy rules required for accepting/publishing those task bundles by some entity that runs a Konflux instance. Another alternative for partner tasks would be to let those to be kept separate and still establish them as a trusted task.